### ISSUE TITLE
no-op as cast

### ISSUE THREAD

**Issue by lolbinarycat** [rust-lang/rust-clippy#15424]
### What it does

Detect when a type is `as` cast as itself.

### Advantage

Removes needless and confusing code.  No-op casts are the only way to cast to non-primitive types, and so it would likely be very confusing to encounter these in the wild.

### Drawbacks

Must take care to not suggest anything that breaks type inference.

### Example

```rust
const FOO: &str = "bar" as &str;
```

Could be written as:

```rust
const FOO: &str = "bar";
```

**Comment by y21** [comment]
Related: there's the `unnecessary_cast` lint that mentions "Checks for casts to the same type" in its description, so it sounds like it should cover this. But it currently actually skips casts to non-primitive types (e.g. if your example was using `i32` then it would get a warning)

https://github.com/rust-lang/rust-clippy/blob/4cb43f4c4128ed1f58d10ade425656caa2a05fa7/clippy_lints/src/casts/unnecessary_cast.rs#L88-L95
